File: MergeSort/src/cpp/src/threadMergeSort.cpp
#include "threadMergeSort.hpp"

ThreadMergeSort::ThreadMergeSort(std::vector<int> *nums) 
    : nums(nums){
}

ThreadMergeSort::~ThreadMergeSort() {}

void ThreadMergeSort::recursiveSort(int left, int right) {

    const int THRESHOLD = 10000;

    if (right - left < THRESHOLD) {
        std::sort(nums->begin() + left, nums->begin() + right + 1);
        return;
    }
    
    if (left >= right) {
        return;
    }
    int mid = left + (right - left) / 2;

    std::thread thread_1([this, left, mid] { this->recursiveSort(left, mid); });
    std::thread thread_2([this, mid, right] { this->recursiveSort(mid + 1, right); });
    thread_1.join();
    thread_2.join();

    std::vector<int> result;
    int i = left;
    int j = mid + 1;

    while (i <= mid && j <= right) {
        if ((*nums)[i] <= (*nums)[j]) {
            result.push_back((*nums)[i]);
            i++;
        } else {
            result.push_back((*nums)[j]);
            j++;
        }
    }

    while (i <= mid) {
        result.push_back((*nums)[i]);
        i++;
    }

    while (j <= right) {
        result.push_back((*nums)[j]);
        j++;
    }

    for (int k = 0; k < result.size(); k++) {
        (*nums)[left + k] = result[k];
    }
}

void ThreadMergeSort::sort() {
    if ((*nums).size() == 0) {
        exit(1);
    }

    std::thread thread_1([this] { this->recursiveSort(0, (*nums).size() - 1); });
    thread_1.join();
}

File: MergeSort/src/Rust/src/thread_merge_sort.rs
use std::thread;

pub struct ParallelMergeSort<'a> {
    nums: &'a mut Vec<i32>,
}

impl<'a> ParallelMergeSort<'a> {
    pub fn new(nums: &'a mut Vec<i32>) -> Self {
        Self { nums }
    }

    pub fn sort(&mut self) {
        self.recursive_sort(0, self.nums.len() - 1);
    }

    fn recursive_sort(&mut self, left: usize, right: usize) {
        const THRESHOLD: usize = 10_000;

        if right - left < THRESHOLD {
            self.nums[left..=right].sort();
            return;
        }

        if left >= right {
            return;
        }

        let mid = left + (right - left) / 2;

        // Clone for parallel threads
        let mut nums_clone_1 = self.nums.clone(); // Declare this as mutable
        let mut nums_clone_2 = self.nums.clone(); // Declare this as mutable

        // Spawn the first thread for the left half
        let thread_1 = thread::spawn(move || {
            let mut sort = ParallelMergeSort::new(&mut nums_clone_1);
            sort.recursive_sort(left, mid);
        });

        // Spawn the second thread for the right half
        let thread_2 = thread::spawn(move || {
            let mut sort = ParallelMergeSort::new(&mut nums_clone_2);
            sort.recursive_sort(mid + 1, right);
        });

        thread_1.join().unwrap();
        thread_2.join().unwrap();

        let mut result = Vec::new();
        let mut i = left;
        let mut j = mid + 1;

        // Merge sorted halves
        while i <= mid && j <= right {
            if self.nums[i] <= self.nums[j] {
                result.push(self.nums[i]);
                i += 1;
            } else {
                result.push(self.nums[j]);
                j += 1;
            }
        }

        while i <= mid {
            result.push(self.nums[i]);
            i += 1;
        }

        while j <= right {
            result.push(self.nums[j]);
            j += 1;
        }

        // Write the merged result back to the original slice
        for (k, &value) in result.iter().enumerate() {
            self.nums[left + k] = value;
        }
    }
}
